src/common/com/intellij/plugins/haxe/ide/folding/HaxeFoldingBuilder.java

```diff
@@ -23,32 +24,116 @@
+  private static final RegionDefinition FD_STYLE = new RegionDefinition("^\\{"+WS+"*region"+WS+"+(.*)$",
+                                                                        "^}"+WS+"?end"+WS+"*region("+WS+"*(.*))?$",
+                                                                        (matcher)->matcher.group(1));
+  private static final RegionDefinition PLUGIN_STYLE = new RegionDefinition("^region"+WS+"*(.*)?$",
+                                                                            "^end"+WS+"*region"+WS+"*(.*)?$", (matcher)->matcher.group(1));
+  private static final RegionDefinition COMMENT_REGIONS[] = { FD_STYLE, PLUGIN_STYLE };
+
+  private static final RegionDefinition CC_REGION = new RegionDefinition("#", "#", null);
+
+
+  private static class RegionDefinition {
+    final Pattern begin;
+    final Pattern end;
+    final Function<Matcher,String> substitutor;
+
+
+    public RegionDefinition(String begin, String end, Function<Matcher, String> titleExtractor) {
+      this.begin = Pattern.compile(begin);
+      this.end = Pattern.compile(end);
+      this.substitutor = titleExtractor;
+    }
+
+    public RegionDefinition(String begin, Function<Matcher, String> titleExtractor) {
+      this.begin = Pattern.compile(begin);
+      this.end = this.begin;
+      this.substitutor = titleExtractor;
+    }
+
+    public Matcher startMatcher(String s) {
+      return begin.matcher(s);
+    }
+
+    public Matcher endMatcher(String s) {
+      return end.matcher(s);
+    }
+
+    public String getPlaceholder(Matcher m) {
+      if (null != substitutor) {
+        return substitutor.apply(m);
+      }
+      return PLACEHOLDER_TEXT;
+    }
+  }
+
+  private static class RegionMarker {
+    public enum Type {
+      START,
+      END,
+      CC,
+      EMPTY
+    }
+
+    public static final RegionMarker EMPTY = new RegionMarker(null, null, null, Type.EMPTY);
+
+    public final RegionDefinition region;
+    public final ASTNode node;
+    public final Type type;
+    public final Matcher matcher;
+
+    public RegionMarker(RegionDefinition region, ASTNode node, Matcher matcher, Type type) {
+      this.region = region;
+      this.node = node;
+      this.type = type;
+      this.matcher = matcher;
+    }
+  }
+
+
+
   @NotNull
   @Override
   public FoldingDescriptor[] buildFoldRegions(@NotNull ASTNode node, @NotNull Document document) {
-    List<FoldingDescriptor> list = new ArrayList<>();
-    buildFolding(node, list);
-    FoldingDescriptor[] descriptors = new FoldingDescriptor[list.size()];
-    return list.toArray(descriptors);
+    List<FoldingDescriptor> descriptorList = new ArrayList<>();
+    List<RegionMarker> regionMarkers = new ArrayList<>();
+    List<RegionMarker> ccMarkers = new ArrayList<>();
+
+    buildFolding(node, descriptorList, regionMarkers, ccMarkers);
+
+    buildMarkerRegions(descriptorList, regionMarkers);
+
+    buildCCMarkerRegions(descriptorList, ccMarkers, document);
```

Comment:
So, maybe this could use a bit of explanation:  buildFolding does the same scan that it did, but for the new region types, it collects the markers into a separate list which is then processed by the two later build functions.  This is for two reasons: the markers don't necessarily follow the tree hierarchy, and they are much harder to process when you only have some of them available.  When they are all collected, it is much easier to create the proper matching logic.